WIP: Polish for publishing

hendraet opened the pull request:
## Description

**Requires a re-install of PROTzilla to get the dependencies for PTM visualization**

This PR fixes a whole bunch of issues and adds features, e.g.:
- Adapts Clustergram so that is actually usable
- Implements PTM visualization tool
- Fixes FormDivider Field and Adds an Info Field
- various little bugs and improvements that were encountered during that journey, e.g.:
  - database settings not being loaded properly under weird circumstances
  - removes protgraph as this utility was not working and will be removed in the future
  - code formatting
- Adds a step (and corresponding workflow) that automatically imports an [example dataset](https://www.ebi.ac.uk/pride/archive/projects/PXD014997). This dataset is currently included in the repo via `git lfs` (which might not be the ideal way since the evidence file is still quite large). 

This PR is based of PR #89.  Since this PR might be obsolete now, the corresponding commits have to be removed from this PR. 

## Testing

- Clustergram can be tested in standard workflow either by using the Tau dataset or the example dataset
- Data for PTM visualization can be under `backend/tests/test_data/ptm_visualization_data`. 
- Some workflows that could serve as a starting point for testing:
  - [PTM visualization](https://github.com/user-attachments/files/23798763/ptm_vis.zip)
  - [PTM visualization with different protein](https://github.com/user-attachments/files/23798762/ptm_vis_GFAB.zip)
  - [Clustergram Test](https://github.com/user-attachments/files/23798765/clustergram_test.zip)

## PR checklist
**Development**
- [ ] If necessary, I have updated the documentation (README, docstrings, etc.)
- [ ] If necessary, I have created / updated tests.
 
**Mergeability**
- [ ] main-branch has been merged into local branch to resolve conflicts
- [ ] The tests and linter have passed AFTER local merge
- [ ] The backend code has been formatted with `black`
- [ ] The frontend code has been formatted with `pnpm format` and checked with `pnpm lint`
 
**Code review**
- [ ] I have self-reviewed my code.
- [ ] At least one other developer reviewed and approved the changes

hendraet committed:
Adds wokring prototype for new Clustergram step

hendraet committed:
Finalizes clustergram plot

hendraet committed:
[WIP] Adds first prototype for ptm visualizations in PROTzilla

hendraet committed:
Improves PTM visualizations (and some formatting)

hendraet committed:
[WIP] Changes PTM vis plots so that they use imported evidence instead of re-importing

hendraet committed:
Improves check for evidence_df + some clean up

hendraet committed:
Fixes FormDivider field and adds extra field for displaying additional information

hendraet committed:
Adds first tests for ptm-visualization

hendraet committed:
Adds tests for PTMs visualizations and provides some minor fixes

hendraet committed:
[WIP] Adds new settings tab for PTM vis and also some utilities to save settings

hendraet committed:
Finishes prototype for PTM visualization setting that also actually uses options from settings

hendraet committed:
Adds default file for PTM settings

hendraet committed:
Further developing the ptm vis settings and doing more finishing touches

hendraet committed:
Adds functioning download button for example_ptm_settings YAML file

hendraet committed:
Shows excluded, but found PTMs in the tables tab

hendraet committed:
Dynamically loads modification grooups from file, adds test, fixes minor issues

hendraet committed:
Minor fixes for PTM vis

hendraet committed:
Reworks and extends tests for ptm-visualizations

hendraet committed:
Final cleanup of PTM visualizations

hendraet committed:
Removes protgraph

hendraet committed:
Fixes bug that leads to databases not being displayed properly in settings

hendraet committed:
Fixes TS lint issues

hendraet committed:
Adds placeholder for example dataset import

hendraet committed:
Fixes test

hendraet committed:
Adapts data table text to indicate that table loading might take a while. Plus some code formatting

hendraet committed:
[WIP] prepares code for new example dataset and fixes bugs surrounding import

hendraet committed:
TODO clean up + code formatting

hendraet committed:
Fixes tests

hendraet committed:
Adds example dataset via git lfs

hendraet committed:
Adds correct evidence.txt

hendraet committed:
Fixes faulty gitignore file and adds missing header-info-filed again

AnnaPolensky commented on backend/protzilla/data_analysis/ptm_visualization/.gitignore:
```diff
@@ -0,0 +1 @@
+tmp
```
Is this some remainder? Or should this still be here?

AnnaPolensky commented on backend/protzilla/data_analysis/ptm_visualization/ptm_bar_plot.py:
```diff
@@ -0,0 +1,64 @@
+def get_bar_plot_config_module(
+    groups_file_path: Path, out_dir: Path
+) -> types.ModuleType:
+    modification_file = out_dir / "result_max_quant_mods.csv"
```
Do we really want to hardcode this path or do we want to extract it to a settings file? Or is this the only type of file that makes sense as an input file?

hendraet replied:
This comes from the underlying [repo](https://github.com/hendraet/ptm-visualization) (which I forked from [here](https://github.com/T4rikA/ptm-visualization)). It's originally a command-line tool that saves intermediate results as csv to disk. This file is a direct consequence of calling the `MaxQuantPreprocessor`.

I tried to change as few things as possible of the original repo as possible, so I decided to keep the intermediate files instead of porting it in a way that puts these results into Python objects and thus into RAM.

AnnaPolensky commented on backend/protzilla/data_analysis/ptm_visualization/ptm_details_plot.py:
```diff
@@ -0,0 +1,116 @@
+def get_details_plot_config_module(
+    groups_file_path: Path, out_dir: Path
+) -> types.ModuleType:
+    modification_file = out_dir / "result_max_quant_mods.csv"
+    cleavage_file = out_dir / "result_max_quant_cleavages.csv"
```
Again, does it make sense to hardcode them here?

hendraet replied:
See above

NeleRiediger commented on backend/protzilla/data_analysis/plots.py:
```diff
@@ -193,34 +200,38 @@ def clustergram_plot(
         sample in `input_df`. Each group will be assigned a color, which will be shown
         in the final plot as a colorbar next to the heatmap. This is an optional
         parameter
     :param flip_axes: If "yes", the rows and columns of the clustergram will be
```
nit: was changed from string to boolean, maybe also change from "yes"/ "no" to true/ false

AnnaPolensky commented on backend/protzilla/utilities/clustergram.py:
```diff
@@ -811,6 +811,9 @@ def _compute_clustered_data(self):
         """
 
+        # 8 is the arbitrary number of colors that is used by the dendrograms (last one is ignored)
+        dendro_colorscale = ["rgb(133,133,133)" for _ in range(8)]
```
Is this really supposed to be eight times the same color?

hendraet replied:
That was a design decision. As you can see in the first plot [here](https://dash.plotly.com/dash-bio/clustergram). The lines of the dendrogram are usually made up of multiple colors, which is something that is not used in figures used in scientific literature. Thus, I decided to unify these colors.

AnnaPolensky commented on frontend/src/components/core/shared/input-fields/header-info-field/header-info-field.tsx:
```diff
@@ -0,0 +1,15 @@
+import { HeaderInfoText } from "../../text";
+
+const DividerContainer = styled.div`
+  //display: inline-block;
```
nit: remove or comment in again?

AnnaPolensky reviewed (changes requested):
Nele and I looked at this commit for the past 2.5 hours. We don't know that much about the codebase yet, so we found it difficult to really sanity check and completely understand everything. Nonetheless we looked at every file and things seemed to make sense apart from our small comments. Still, for both of us, Protzilla is not working properly if we check out your branch?

I cannot add any runs anymore:
<img width="1895" height="710" alt="Image" src="https://github.com/user-attachments/assets/b5aa1c2e-3184-4e62-a962-99e661536cb0" />

Nele only gets an error when trying to open Protzilla:
![Image](https://github.com/user-attachments/assets/61a78879-3332-4d47-aa2a-18d64e61f237)

Next time it would be great if you could do separate PRs for separate issues and not put the formatter commit that is touching a bunch of files you never touched into the same PR as your changes.

hendraet commented:
Thank you for the comments, I will have a look at it in the next couple of days. :) 

I totally agree that running the formatting on all files might not be the best idea regards formatting. I also reverted it in the PR #106 to make that review there easier.

Regarding Nele's issue, I thought I had fixed it in the latest commit (from 20.11.). Has she already checked it out?
Unfortunately, I have no clue what causes Anna's issue (classic: works on my machine). Maybe we should have a quick meeting and try to resolve it.

hendraet committed:
Reverts formatting of untouched files

hendraet commented:
Updated the description so that it also mentions the re-installation of dependencies that should resolve the startup issues.

I also tried to revert most  of the code formatting changes of untouched files (at least removed ~50 files from the changelog. Hope that helps a bit)

Also, added three workflows that could be valuable for testing.

hendraet committed:
Addresses minor issues mentioned in review

hendraet committed:
Merge remote-tracking branch 'origin/dev' into polish-for-publishing

tE3m commented on protzilla_dev.sh:
```diff
@@ -40,7 +40,7 @@ trap cleanup EXIT INT TERM
 cd frontend
 pnpm install
-pnpm build
+#pnpm build
```
nitpick: why not remove the line instead of commenting it out?

tE3m commented on frontend/src/components/app/settings/other-settings/ptm-vis-settings.tsx:
For color groups without values, explicitly write "no colors defined" in the HTML (else we have just a whitespace or EOT)

hendraet replied:
Added a check that this can't happen. If colors are not defined, this will also lead to errors when actually plotting

tE3m commented on backend/protzilla/methods/data_analysis.py:
```diff
@@ -962,36 +956,59 @@ def insert_dataframes(self, steps: StepManager, inputs) -> dict:
+            form["metadata_column"].options = form_helper.get_choices_for_metadata_non_sample_columns(
+                run,
+                instance_identifier=form.values['metadata_df']
+            )
```
Maybe use set_options() instead of .options = form... Check how it is done in other forms on dev.

tE3m commented on backend/protzilla/methods/data_analysis.py:
```diff
@@ -2218,3 +2090,103 @@ def insert_dataframes(self, steps: StepManager, inputs) -> dict:
+    def create_form(self):
+        base_form = super(_PTMVisualizationWithGroups, self).create_form()
+        form = Form(
+            label=base_form.label,
```
should be overwrittten by the individual visualisation forms

hendraet replied:
I have refactored it, but I was wondering if this label is actually used somewhere? At least, my IDE was not able to show me any usages. Do you see any, or could this be removed safely?

tE3m commented on frontend/src/components/app/settings/other-settings/ptm-vis-settings.tsx:
```diff
@@ -0,0 +1,337 @@
+    <div>
+      <Text text={name + ":"} style={{ fontWeight: "bold" }} />
+      <Text text=" Color: " />
+      <ColorText text={color} />
```
As the project locale is EN_GB, all user-facing text should conform to UK English. Here, it should be "colour"

tE3m commented on backend/protzilla/importing/peptide_import.py:
```diff
@@ -68,37 +55,40 @@ def peptide_import(file_path : Path, intensity_name, map_to_uniprot) -> dict:
         low_memory=False,
         na_values=["", 0],
         keep_default_na=True,
+        usecols=lambda x: (x.capitalize() if ' ' in x else x) in id_columns,
```
Not very legible... Maybe explicitly define the function since we need it twice?

hendraet replied:
Do you like the new version better?

tE3m commented on backend/protzilla/importing/peptide_import.py:
```diff
@@ -4,45 +4,32 @@
-    peptide_intensity_name = (
-        "LFQ intensity" if intensity_name == "iBAQ" else intensity_name
-    )
+    peptide_intensity_name = IntensityType.INTENSITY.value
```
why is this hardcoded? Why are the values defined in the enum if we only ever allow "Intensity"?

hendraet replied:
Peptides only have an intensity value, while proteins can also be quantified via iBAQ or LFQ intensity. I added a comment for clarificaiton

hendraet replied:
This enum is used for proteins and peptides. However, in peptide files, we only ever have Intensity, while iBAQ/LFQ Intensity is only used in protein files. One could, of course, argue that we should rather use a single string (if it cannot assume multiple values yet) or that we should use a different enum for peptides. 
Any preferences?

tE3m commented on backend/protzilla/data_analysis/ptm_visualization/ptm_vis_utils.py:
```diff
@@ -0,0 +1,159 @@
+        'OUTPUT_FOLDER': out_dir,
+
+        # Plot Settings
+        # 0 for horizontal, 1 for vertical, note figure height and width are then automatically swapped
```
Why not code this directly? Useless if/else overhead

hendraet replied:
Unfortunately, the underlying visualization package handles it this way. I personally would have handled it differently, but I didn't want to refactor the underlying code too much. (It has already cost me a lot of time to port it from a pure command-line tool to sth. that could be included in PROTzilla).

tE3m commented on backend/protzilla/data_analysis/plots.py:
```diff
@@ -193,34 +196,36 @@ def clustergram_plot(
+            assert metadata_column in metadata_df.columns
+            # TODO: debatable if this filtering should be done here or in the filtering steps
+            filtered_metadata_df = metadata_df[metadata_df['Sample'].isin(input_df_wide.index)]
+
```
What does the filtering accomplish and why was the additional assertion removed?

hendraet replied:
When I encountered the assertion error (because of a mismatch of protein and metadata file), I thought that the assertion error was not expressive enough. 
I fixed it this way because I thought that it is just more convenient to the user if it is handled like this, e.g., if the user wants to process only a subset of their data, they only have to modify the `protein_groups.txt` and the metadata file can stay the same.  
Of course, one could also add a descriptive string to the assertion or raise a proper error. 

I removed the second assertion because I thought that it would be obsolete after the filtering, because it would be included in the length check. (If we have additional samples in metadata, they will be filtered out, and if we have fewer, the length check should fail. Or am I mistaken?)

If we want to keep the filtering instead of throwing an error, I would be interested in your opinion on whether this should be done here or somewhere else.

tE3m commented on backend/protzilla/data_analysis/plots.py:
```diff
@@ -262,24 +267,20 @@ def clustergram_plot(
-        ) != sorted(sample_group_df.index.values.tolist()):
+        elif isinstance(metadata_df, pd.DataFrame) and metadata_column not in metadata_df.columns:
+            msg = "The column selected for annotation is not present in the corresponding metadata dataframe"
+        elif isinstance(metadata_df, pd.DataFrame) and len(input_df_wide) != len(filtered_metadata_df):
```
see above

hendraet replied:
I'm sorry, but what exactly is this referencing? :grimacing:

tE3m commented on .gitignore:
```diff
@@ -110,4 +110,7 @@ venv.bak/
\ No newline at end of file
+.pyre/
+
+/src/
```
what does this do? why?

hendraet replied:
This is to exclude the files of the PTM visualization/`protein_sequencing` repo that are placed here because they are installed in "editable" mode.

(Why are they installed in editable mode? Excellent question. The repo calls a binary file to do calculations, and I didn't find a way to include this binary without calling by giving a relative path. You can check the code [here](https://github.com/hendraet/ptm-visualization/blob/a972aaa5a5ea50df859cf4ddd0e3d16c63e19cb4/protein_sequencing/uniprot_align.py#L30-L39). I'm also happy, about any suggestions here)

tE3m commented:
Regarding the PTM visualisation: We weren't able to test the added functionality since the file paths in the run are the (absolute) ones from your filesystem :)

hendraet commented:
> Regarding the PTM visualisation: We weren't able to test the added functionality since the file paths in the run are the (absolute) ones from your filesystem :)

Oh man, I was kinda expecting that the workflow import hasn't been tested properly in a long time and things would break. But I guess it's a good thing that you tested it again 😅 
I will send you more information regarding the testing of these steps in the next couple of days :)

hendraet committed:
Fixes some more review comments

hendraet reviewed (commented):
Thanks for the feedback. :) 
I tried to answer the questions and already implemented some changes, but due to time constraints, I was not able to address all issues already

hendraet commented:
> Regarding the PTM visualisation: We weren't able to test the added functionality since the file paths in the run are the (absolute) ones from your filesystem :)

To test the visualizations, you can start with the files provided as test data (`backend/tests/test_data/ptm_visualization_data`). There's one subfolder (`P10636`) that has all the files for the Tau protein, while the main directory contains files for GFAB. The files are named in a way so that they are hopefully easy to match to respective form field. The filenames for the "Happy Path" are always quite short, while longer names usually indicate what kind of error they represent.  

In case you also cannot see the steps of the workflow, you just need an Evidence Import step and can then add the steps for the PTM visualization.

Hope that serves as a starting point for testing :)

jorisfu commented:
Appears to be working. The plot download looks a bit messed up though (see attached image). Anything left to do before we can merge? (apart from merge/rebase to current dev)

<img width="1350" height="741" alt="mynd" src="https://github.com/user-attachments/assets/acc5720f-c933-48b7-9ec1-9b19a09dda5c" />

hendraet committed:
Handles upload of malformed PTM settings and adds tests for it

hendraet committed:
Merge remote-tracking branch 'origin/dev' into polish-for-publishing

hendraet committed:
Formats code

hendraet committed:
Fixes faulty test case

hendraet committed:
Skips non-working test
